File: mobile/ios/BroadcastExtension/UdpAudioSender.swift
// UdpAudioSender.swift
//
// 构建 AudioPacket（32B 大端头 + ChaCha20-Poly1305 密文‖tag）并 UDP 发送。
// 与 desktop/src-tauri/src/network/packet.rs 字节级互通（spec §2）。
// 加密用 CryptoKit ChaChaPoly；UDP 用 BSD socket（轻量，适合 Extension）。

import Foundation
import CryptoKit
import Darwin

final class UdpAudioSender {
    private let config: SessionConfig
    private let key: SymmetricKey
    private var sequence: UInt32 = 0
    private var timestamp: UInt64 = 0
    private var sock: Int32 = -1
    private var dest: sockaddr_in = sockaddr_in()

    /// 常量（对齐 constants）。
    private let magic: UInt16 = 0x534C
    private let version: UInt8 = 1
    private let headerLen: UInt8 = 32
    private let codec: UInt8 = 1
    private let flagStreamEnd: UInt8 = 0x01

    init?(config: SessionConfig) {
        self.config = config
        let keyBytes = config.audioKeyBytes()
        guard keyBytes.count == 32 else { return nil }
        self.key = SymmetricKey(data: keyBytes)

        sock = socket(AF_INET, SOCK_DGRAM, 0)
        guard sock >= 0 else { return nil }
        dest.sin_family = sa_family_t(AF_INET)
        dest.sin_port = in_port_t(config.audioPort).bigEndian
        dest.sin_addr.s_addr = inet_addr(config.targetHost)
        if dest.sin_addr.s_addr == INADDR_NONE { return nil }
    }

    deinit {
        if sock >= 0 { close(sock) }
    }

    /// 加密并发送一帧 Opus 数据。返回是否成功。
    @discardableResult
    func send(opusFrame: Data, streamEnd: Bool = false) -> Bool {
        guard sock >= 0 else { return false }
        let header = buildHeader(
            payloadLen: UInt16(opusFrame.count),
            flags: streamEnd ? flagStreamEnd : 0)

        // nonce = stream_id(4 BE) ‖ sequence(4 BE) ‖ 0(4)
        let nonceBytes = buildNonce()
        guard let nonce = try? ChaChaPoly.Nonce(data: nonceBytes) else { return false }
        guard let sealed = try? ChaChaPoly.seal(
            opusFrame,
            using: key,
            nonce: nonce,
            authenticating: header) else {
            return false
        }

        // packet = header ‖ ciphertext ‖ tag
        var packet = header
        packet.append(sealed.ciphertext)
        packet.append(sealed.tag)

        let ok = packet.withUnsafeBytes { (raw: UnsafeRawBufferPointer) -> Bool in
            guard let base = raw.baseAddress else { return false }
            let sent = withUnsafePointer(to: &dest) { dPtr -> Int in
                dPtr.withMemoryRebound(to: sockaddr.self, capacity: 1) { sa in
                    sendto(sock, base, packet.count, 0, sa, socklen_t(MemoryLayout<sockaddr_in>.size))
                }
            }
            return sent == packet.count
        }

        sequence &+= 1
        timestamp &+= UInt64(config.sampleRate / 1000 * config.frameDurationMs) // +480
        return ok
    }

    /// 构建 32 字节大端头部。
    private func buildHeader(payloadLen: UInt16, flags: UInt8) -> Data {
        var d = Data(count: Int(headerLen))
        d[0] = UInt8(magic >> 8)
        d[1] = UInt8(magic & 0xFF)
        d[2] = version
        d[3] = headerLen
        writeBE32(&d, at: 4, UInt32(config.streamId))
        writeBE32(&d, at: 8, sequence)
        writeBE64(&d, at: 12, timestamp)
        d[20] = codec
        d[21] = UInt8(config.channels)
        d[22] = UInt8(config.frameDurationMs)
        d[23] = flags
        writeBE32(&d, at: 24, UInt32(config.sampleRate))
        d[28] = UInt8(payloadLen >> 8)
        d[29] = UInt8(payloadLen & 0xFF)
        d[30] = 0; d[31] = 0
        return d
    }

    private func buildNonce() -> Data {
        var n = Data(count: 12)
        writeBE32(&n, at: 0, UInt32(config.streamId))
        writeBE32(&n, at: 4, sequence)
        // 8..12 = 0
        return n
    }

    private func writeBE32(_ d: inout Data, at offset: Int, _ v: UInt32) {
        d[offset] = UInt8((v >> 24) & 0xFF)
        d[offset + 1] = UInt8((v >> 16) & 0xFF)
        d[offset + 2] = UInt8((v >> 8) & 0xFF)
        d[offset + 3] = UInt8(v & 0xFF)
    }

    private func writeBE64(_ d: inout Data, at offset: Int, _ v: UInt64) {
        for i in 0..<8 {
            d[offset + i] = UInt8((v >> ((7 - i) * 8)) & 0xFF)
        }
    }
}

File: mobile/flutter_app/android/app/src/main/kotlin/com/soundlink/soundlink/network/UdpAudioSender.kt
// UdpAudioSender.kt
//
// 构建 AudioPacket（32B 大端头 + ChaCha20-Poly1305 密文‖tag）并 UDP 发送。
// 与 desktop/src-tauri/src/network/packet.rs 字节级互通（spec §2）。
// 加密用 BouncyCastle ChaCha20-Poly1305（允许自定义 12B nonce）。

package com.soundlink.soundlink.network

import org.bouncycastle.crypto.engines.ChaCha20Poly1305
import org.bouncycastle.crypto.macs.AEADMac
import org.bouncycastle.crypto.params.AEADParameters
import org.bouncycastle.crypto.params.KeyParameter
import java.net.DatagramPacket
import java.net.DatagramSocket
import java.net.InetAddress
import java.nio.ByteBuffer
import java.nio.ByteOrder

data class SenderConfig(
    val targetHost: String,
    val audioPort: Int,
    val streamId: Int,
    val audioKey: ByteArray,      // 32B
    val sampleRate: Int,
    val channels: Int,
    val frameDurationMs: Int,
    val bitrate: Int,
)

class UdpAudioSender(private val cfg: SenderConfig) : AutoCloseable {

    private val key = KeyParameter(cfg.audioKey)
    private var sequence = 0
    private var timestamp = 0L
    private val socket = DatagramSocket()
    private val dest = InetAddress.getByName(cfg.targetHost)

    @Synchronized
    fun send(opusFrame: ByteArray, streamEnd: Boolean = false): Boolean {
        val header = buildHeader(opusFrame.size, if (streamEnd) FLAG_STREAM_END else 0)
        val cipher = encrypt(opusFrame, header)

        val packet = ByteArray(header.size + cipher.size)
        System.arraycopy(header, 0, packet, 0, header.size)
        System.arraycopy(cipher, 0, packet, header.size, cipher.size)

        return try {
            socket.send(DatagramPacket(packet, packet.size, dest, cfg.audioPort))
            sequence++
            timestamp += (cfg.sampleRate / 1000 * cfg.frameDurationMs).toLong() // +480
            true
        } catch (e: Exception) {
            false
        }
    }

    /** 构建 32 字节大端头部。 */
    private fun buildHeader(payloadLen: Int, flags: Int): ByteArray {
        val bb = ByteBuffer.allocate(HEADER_LEN).order(ByteOrder.BIG_ENDIAN)
        bb.putShort(MAGIC)              // 0..2 magic
        bb.put(VERSION.toByte())        // 2 version
        bb.put(HEADER_LEN.toByte())     // 3 header_len
        bb.putInt(cfg.streamId)         // 4..8 stream_id
        bb.putInt(sequence)             // 8..12 sequence
        bb.putLong(timestamp)           // 12..20 timestamp
        bb.put(CODEC_OPUS.toByte())     // 20 codec
        bb.put(cfg.channels.toByte())   // 21 channels
        bb.put(cfg.frameDurationMs.toByte()) // 22 frame_duration_ms
        bb.put(flags.toByte())          // 23 flags
        bb.putInt(cfg.sampleRate)       // 24..28 sample_rate
        bb.putShort(payloadLen.toShort()) // 28..30 payload_len
        bb.putShort(0)                  // 30..32 reserved
        return bb.array()
    }

    /** nonce = stream_id(4 BE) ‖ sequence(4 BE) ‖ 0(4)。 */
    private fun buildNonce(): ByteArray {
        val bb = ByteBuffer.allocate(12).order(ByteOrder.BIG_ENDIAN)
        bb.putInt(cfg.streamId)
        bb.putInt(sequence)
        bb.putInt(0)
        return bb.array()
    }

    /** ChaCha20-Poly1305 AEAD，返回 ciphertext ‖ tag(16B)。 */
    private fun encrypt(plaintext: ByteArray, aad: ByteArray): ByteArray {
        val engine = ChaCha20Poly1305()
        engine.init(true, AEADParameters(key, 128, buildNonce(), aad))
        val out = ByteArray(plaintext.size + 16)
        val len = engine.processBytes(plaintext, 0, plaintext.size, out, 0)
        engine.doFinal(out, len)
        return out
    }

    override fun close() {
        socket.close()
    }

    companion object {
        const val MAGIC: Short = 0x534C
        const val VERSION = 1
        const val HEADER_LEN = 32
        const val CODEC_OPUS = 1
        const val FLAG_STREAM_END = 0x01
    }
}
